rig/machine_control/regions.py

```diff
@@ -0,0 +1,200 @@
+"""SpiNNaker machine regions.
+
+.. todo::
+    Neaten this documentation!
+
+Regions (not to be confused with rig Regions) are used to specify areas of a
+SpiNNaker machine for the purposes of transmitting nearest neighbour packets or
+for determining which chips should be included in any flood-fill of data or
+application loading.
+
+Regions are split into 4 levels (0, 1, 2 and 3) referring to increasing
+coarseness in the number of chips they are able to represent.  Level 3 splits a
+256x256 SpiNNaker machine (the largest possible) into a grid of m nxn chunks,
+in each chunk blocks of pxp chips may be selected.  Level 2 splits each nxn
+chunk into m oxo chunks, and so on.  By level 3 it is possible to select
+individual chips from a 4x4 grid of chips.
+
+A 32-bit value representing a region uses the top 16 bits (31:16) to represent
+the x- and y-coordinates of the region and the level and the lower 16 bits
+(15:0) to represent which of the 16 blocks contained within the chunk should be
+selected.
+"""
+import collections
+from six import iteritems
+
+
+def get_region_for_chip(x, y, level=3):
+    """Get the region word for the given chip co-ordinates.
+
+    Parameters
+    ----------
+    x : int
+        x co-ordinate
+    y : int
+        y co-ordinate
+    level : int
+        Level of region to build. 0 is the most coarse and 3 is the finest.
+        When 3 is used the specified region will ONLY select the given chip,
+        for other regions surrounding chips will also be selected.
+
+    Returns
+    -------
+    int
+        A 32-bit value representing the co-ordinates of the chunk of SpiNNaker
+        chips that should be selected and the blocks within this chunk that are
+        selected.  As long as bits (31:16) are the same these values may be
+        OR-ed together to increase the number of sub-blocks selected.
+    """
+    shift = 6 - 2*level
+
+    bit = ((x >> shift) & 3) + 4*((y >> shift) & 3)  # bit in bits 15:0 to set
+
+    mask = 0xffff ^ ((4 << shift) - 1)  # in {0xfffc, 0xfff0, 0xffc0, 0xff00}
+    nx = x & mask  # The mask guarantees that bits 1:0 will be cleared
+    ny = y & mask  # The mask guarantees that bits 1:0 will be cleared
+
+    #        sig bits x | sig bits y |  2-bit level  | region select bits
+    region = (nx << 24) | (ny << 16) | (level << 16) | (1 << bit)
+    return region
+
+
+def reduce_regions_hierarchically(chips):
```

Comment:
This could safely be called `coalesce_regions` or `minimise_regions` since the implementation is immaterial and the effect is a perfect minimisation of the set of regions required to address exactly the given collection of chips.